### PR TITLE
Add support for Linux in HTTP client (close #5)

### LINKED ISSUE #5: Add support for Linux in HTTP client

### PR DESCRIPTION
**Pull request by matus-tomlein**
This PR refactors the HTTP clients and adds an implementation for Linux using Curl (issue #5). There are the following changes:

* Separate out the Windows, and Apple implementations to separate classes (`HttpClientWindows`, `HttpClientApple`) that implement an abstract base class `HttpClient`
* Move HTTP client-related classes to a `http` subfolder (similar to Java tracker)
* Enable passing a custom `HttpClient` as the last optional argument in the `Emitter` constructor. It is passed as a [C++ unique pointer](https://docs.microsoft.com/en-us/cpp/cpp/smart-pointers-modern-cpp?view=msvc-170#c-standard-library-smart-pointers) which means that the Emitter will take ownership of the pointer and it will be deleted when the Emitter instance is deleted. My assumption is that HTTP clients belong to Emitters.
* Add a `HttpClientCurl` implementation of `HttpClient` that uses the easy interface in the Curl library. It is used by default on non-Windows and non-Apple systems.

So far, we don't have integration tests in the tracker, so I didn't add any tests for the new Curl client. It can be tested using the example app by running `./build/example/tracker_example 127.0.0.1:9090` after `make`. I want to add integration tests using Micro and CI on Linux in other issues so that this PR doesn't get too bloated.

I also made a [PR with some small docs updates on data-value-resources](https://github.com/snowplow-incubator/data-value-resources/pull/74/files).

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -18,9 +18,9 @@ cc-objects := $(patsubst %.c, %.o, $(cc-include-files))
 
 # C++ Files
 
-cxx-src-files := $(shell find src -maxdepth 1 -name "*.cpp")
+cxx-src-files := $(shell find src -maxdepth 2 -name "*.cpp")
```
Comment (AlexBenny) [inline Makefile:21]: I guess it's my lack of knowledge in the C++ world but I'm wondering why we added "maxdepth". Is it a best practice? Could we have subfolders with files that we don't want to import?

File: src/http/http_client_curl.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+const string HttpClientCurl::TRACKER_AGENT = string("Snowplow C++ Tracker (Unix)");
+
+static size_t write_data(void *data, size_t byte_size, size_t n_bytes, std::string *body) {
+  return byte_size * n_bytes;
+}
+
+HttpRequestResult HttpClientCurl::http_request(const RequestMethod method, CrackedUrl url, const string &query_string, const string &post_data, list<int> row_ids, bool oversize) {
+  CURL *curl = curl_easy_init();
+
+  if (curl) {
```
Comment (AlexBenny) [inline src/http/http_client_curl.cpp]: Just a very nit opinion: maybe we can return the failure option early?

File: test/http/http_client_test.cpp [hunk 3]
```diff
@@ -11,62 +11,58 @@ software distributed under the Apache License Version 2.0 is distributed on an
...
 */
 
-#include "../src/http_client.hpp"
-#include "catch.hpp"
+#include "test_http_client.hpp"
+#include "../catch.hpp"
 
 using namespace snowplow;
 
-#if defined(SNOWPLOW_TEST_SUITE)
```
Comment (AlexBenny) [inline test/http/http_client_test.cpp:19]: Awesome!! Thanks! 🎉  Also, with can close the issue #12 !!

**Review by AlexBenny — APPROVED** [review]
Great job, it looks good!! 👍 
...even if C++ isn't the nicest language to read! :)

Just a couple of notes about the doc part:
- I guess we should update the README too. At the moment it reports just Mac and Windows in the Quick Start section.
- Same for the documentation, the [setup guide](https://github.com/snowplow-incubator/data-value-resources/blob/main/50%20Documentation/cpp_tracker/01-setup.md) reports just Mac and Win.

File: Makefile (hunk 1 above)
Comment (matus-tomlein) [reply Makefile:21]: Not sure what is the point of it, but C++ produces a lot of garbage during build in the `src` folder. I don't see it making any deep folder structure, but maybe this is just being careful. I do want to separate the build output to a different folder, will adress it in a future issue.

File: src/http/http_client_curl.cpp (hunk 2 above)
Comment (matus-tomlein) [reply src/http/http_client_curl.cpp]: Makes sense, changed!

File: test/http/http_client_test.cpp (hunk 3 above)
Comment (matus-tomlein) [reply test/http/http_client_test.cpp:19]: That's great, didn't realize there was a task for it!

**Comment by matus-tomlein** [comment]
Thanks @AlexBenny! The Linux support is not yet complete, there still remain some other places to add support (desktop context, CI). I will update the documentation at once in issue #50

File: test/http/http_client_test.cpp (hunk 3 above)
Comment (AlexBenny) [reply test/http/http_client_test.cpp:19]: At least my issue-mapping has been good for something ;)

**Pull request merged**